plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftLaunchController.java

```diff
@@ -104,109 +125,140 @@ private String getLabel(String mode) {
+		}
+		return false;
 	}
-	
 
-	protected boolean waitForDeploymentConfigAndPods(IServer server, IProgressMonitor monitor) {
-		boolean podsReady = waitForPodsReady(server, monitor);
-		if (podsReady && !monitor.isCanceled()) {
-			return waitForDeploymentConfigReady(server, monitor);
+	private boolean waitForPodsReady(IServer server, IProgressMonitor monitor) {
+		while (!monitor.isCanceled()) {
+			Collection<IPod> pods = OpenShiftServerUtils.getAllPods(server); // result is possibly null
+            Collection<IPod> buildPods = ResourceUtils.getBuildPods(pods);
+            Collection<IPod> other = ResourceUtils.getRuntimePods(pods);
+            if (!areComplete(buildPods, other)) {
+                sleep(RECHECK_DELAY);
+            } else {
+                return true;
+            } 
 		}
 		return false;
 	}
 
-	private OpenShiftServerBehaviour getOpenShiftServerBehaviour(ILaunchConfiguration configuration)
-			throws CoreException {
-		IControllableServerBehavior serverBehavior = getServerBehavior(configuration);
-		if (!(serverBehavior instanceof OpenShiftServerBehaviour)) {
-			throw toCoreException("Unable to find a OpenShiftServerBehaviour instance");
+	private boolean areComplete(Collection<IPod> buildPods, Collection<IPod> other) {
+		if (CollectionUtils.isEmpty(buildPods) 
+				|| CollectionUtils.isEmpty(other)) {
+			return false;
 		}
-		return (OpenShiftServerBehaviour) serverBehavior;
+		if (!areAllComplete(buildPods)) {
+			return false; // wait longer
+		}
+		if (!areAllRunning(other)) {
+			return false; // wait longer
+		}
+
+		return true;
 	}
 
-	private boolean waitForDeploymentConfigReady(IServer server, IProgressMonitor monitor) {
-		while (!monitor.isCanceled()) {
-			try {
-				OpenShiftServerUtils.getReplicationController(server);
-				return true;
-			} catch(CoreException ce) {
-				sleep(RETRY_DELAY);
-			}
-		}
-		return false;
+	private boolean areAllComplete(Collection<IPod> pods) {
+		return areAllRunning(pods, false);
 	}
-	
-	private boolean waitForPodsReady(IServer server, IProgressMonitor monitor) {
-		while (!monitor.isCanceled()) {
-			IPod[] pods = findPods(server); // result is possibly null
-			if (pods != null) {
-                IPod[] buildPods = findBuildPods(pods);
-                IPod[] other = findRunnablePods(pods);
-                if (!complete(buildPods, other)) {
-                    sleep(RETRY_DELAY);
-                } else {
-                    return true;
-                } 
-            } else {
-                return false;
-            }
+
+	private boolean areAllRunning(Collection<IPod> pods) {
+		return areAllRunning(pods, true);
+	}
+
+	private boolean areAllRunning(Collection<IPod> pods, boolean expectedRunning) {
+		for (Iterator<IPod> it = pods.iterator(); it.hasNext(); ) {
+			boolean running = POD_STATE_RUNNING.equalsIgnoreCase(it.next().getStatus());
+			if (expectedRunning == !running) {
+				return false;
+			} else if (expectedRunning == running) {
```

Comment:
lol, indeed +100.  this looks like a snippet worth publishing to "coding horror".